File: haskell/projects/calculator.hs
module Main where
-- PROJECT: Pure Functional Calculator | runghc calculator.hs
-- Docs: https://www.haskell.org/

-- TODO 1 (Mini): getLine REPL, words to parse "3 + 4"
-- main :: IO ()
-- main = loop
--   where loop = do
--           putStr ">> "
--           line <- getLine
--           if line == "exit" then return ()
--           else do
--             case words line of
--               [a, op, b] -> case (reads a, reads b) of
--                 ([(x,"")] , [(y,"")]) -> case op of
--                   "+" -> print (x+y :: Double)
--                   "-" -> print (x-y)
--                   "*" -> print (x*y)
--                   "/" | y == 0    -> putStrLn "Error: division by zero"
--                       | otherwise -> print (x/y)
--                   _   -> putStrLn $ "Unknown operator: " ++ op
--                 _ -> putStrLn "Invalid numbers"
--               _ -> putStrLn "Usage: number op number"
--             loop

-- TODO 2 (Intermediate): Parser combinators
-- newtype Parser a = Parser { runParser :: String -> Maybe (a, String) }
-- instance Functor     Parser where ...
-- instance Applicative Parser where ...
-- instance Monad       Parser where ...
-- instance Alternative Parser where ...
--
-- charP :: Char -> Parser Char
-- digitP :: Parser Char
-- numberP :: Parser Double
-- opP :: Parser Char
--
-- -- Combine: exprP = termP `chainl1` addOp
-- -- where addOp = (+) <$ charP '+' <|> (-) <$ charP '-'
-- --       termP = factorP `chainl1` mulOp

-- TODO 3 (Advanced): AST with full evaluation
-- data Expr = Num Double
--           | BinOp Char Expr Expr
--           | Neg Expr
--           deriving (Show)
--
-- eval :: Expr -> Either String Double
-- eval (Num n)       = Right n
-- eval (Neg e)       = fmap negate (eval e)
-- eval (BinOp '+' a b) = (+) <$> eval a <*> eval b
-- eval (BinOp '-' a b) = (-) <$> eval a <*> eval b
-- eval (BinOp '*' a b) = (*) <$> eval a <*> eval b
-- eval (BinOp '/' a b) = do
--   bv <- eval b
--   if bv == 0 then Left "Division by zero" else fmap (/bv) (eval a)

main :: IO ()
main = putStrLn "TODO: implement calculator"

File: ocaml/projects/calculator.ml
(* PROJECT: OCaml Calculator | ocaml calculator.ml *)

(* TODO 1 (Mini): REPL with simple expression parsing
   let rec repl () =
     print_string ">> ";
     match input_line stdin with
     | exception End_of_file -> ()
     | "exit" -> ()
     | line ->
       (try
         let tokens = String.split_on_char ' ' (String.trim line) in
         match tokens with
         | [a; op; b] ->
           let a = float_of_string a and b = float_of_string b in
           let result = match op with
             | "+" -> a +. b | "-" -> a -. b
             | "*" -> a *. b | "/" -> if b = 0.0 then raise Division_by_zero else a /. b
             | _ -> failwith ("Unknown op: " ^ op)
           in Printf.printf "%g\n" result
         | _ -> print_endline "Usage: a op b"
       with exn -> Printf.printf "Error: %s\n" (Printexc.to_string exn));
       repl ()  *)

(* TODO 2 (Intermediate): Recursive descent parser with ADT
   type expr =
     | Num of float
     | Add of expr * expr
     | Sub of expr * expr
     | Mul of expr * expr
     | Div of expr * expr
     | Neg of expr

   (* Tokenizer: string -> string list *)
   (* parse_expr, parse_term, parse_factor *)
   (* eval: expr -> float *)  *)

(* TODO 3 (Advanced): Variables via Hashtbl
   let vars : (string, float) Hashtbl.t = Hashtbl.create 16
   (* Parse "let x = 5 + 3" → Hashtbl.replace vars "x" (eval expr) *)
   (* Reference variable: look up in vars table *)  *)

let () =
  print_endline "TODO: implement OCaml calculator";
  (* Uncomment: repl () *)
